### PR TITLE
Improve dotnet SDKs installation

### PR DESCRIPTION
**Pull request by LeonidLapshin**
# Description
Dotnet SDKs installation process include "curl" that needs to be replaced by "download_with_retry" function to make installation process less flaky due to network errors/timeouts.

## Check list
- [ ] Related issue / work item is attached
- [ ] Tests are written (if applicable)
- [ ] Documentation is updated (if applicable)
- [ ] Changes are tested and related VM images are successfully generated

### CODE REVIEW

File: images/linux/scripts/installers/dotnetcore-sdk.sh [hunk 1]
```diff
@@ -46,34 +46,33 @@ done
...
 
+# Download additional SDKs in parallel
 for sdk in $sortedSdks; do
-    url="https://dotnetcli.blob.core.windows.net/dotnet/Sdk/$sdk/dotnet-sdk-$sdk-linux-x64.tar.gz"
-    echo "$url" >> urls
-    echo "Adding $url to list to download later"
+    (url="https://dotnetcli.blob.core.windows.net/dotnet/Sdk/$sdk/dotnet-sdk-$sdk-linux-x64.tar.gz"
+    download_with_retries $url) &
 done
+wait
```
Comment (maxim-lobanov) [inline images/linux/scripts/installers/dotnetcore-sdk.sh:63]: Just a suggestion: did you consider to parallel installation via `parallel` tool?
https://www.gnu.org/software/bash/manual/html_node/GNU-Parallel.html

Comment (LeonidLapshin) [reply images/linux/scripts/installers/dotnetcore-sdk.sh:63]: I guess it do mostly the same but include one more tool, while current option provide background processes through OS core features

File: images/linux/scripts/installers/dotnetcore-sdk.sh [hunk 2]
```diff
@@ -46,34 +46,33 @@ done
...
 done
 
 sortedSdks=$(echo ${sdks[@]} | tr ' ' '\n' | grep -v preview | grep -v rc | grep -v display | cut -d\" -f2 | sort -u -r)
 
+# Download additional SDKs in parallel
 for sdk in $sortedSdks; do
-    url="https://dotnetcli.blob.core.windows.net/dotnet/Sdk/$sdk/dotnet-sdk-$sdk-linux-x64.tar.gz"
-    echo "$url" >> urls
-    echo "Adding $url to list to download later"
+    (url="https://dotnetcli.blob.core.windows.net/dotnet/Sdk/$sdk/dotnet-sdk-$sdk-linux-x64.tar.gz"
```
Comment (al-cheb) [inline images/linux/scripts/installers/dotnetcore-sdk.sh:60]: Why do we need ()?

File: images/linux/scripts/installers/dotnetcore-sdk.sh [hunk 3]
```diff
@@ -46,34 +46,33 @@ done
...
+    download_with_retries $url) &
```
Comment (al-cheb) [inline images/linux/scripts/installers/dotnetcore-sdk.sh:61]: Where do you check exitcode of the jobs?

File: images/linux/scripts/installers/dotnetcore-sdk.sh (hunk 2 above)
Comment (LeonidLapshin) [reply images/linux/scripts/installers/dotnetcore-sdk.sh:60]: at the end of the next line there is ") &" to make processes run in background

Comment (al-cheb) [reply images/linux/scripts/installers/dotnetcore-sdk.sh:60]: Why can't we `download_with_retries $url &` use?

File: images/linux/scripts/installers/dotnetcore-sdk.sh (hunk 3 above)
Comment (LeonidLapshin) [reply images/linux/scripts/installers/dotnetcore-sdk.sh:61]: there is no such functionality right now, I probably will use gnu parallel tool as @maxim-lobanov mentioned as it has child error handling build-in, thank you for mentioning this!

File: images/linux/scripts/installers/dotnetcore-sdk.sh (hunk 1 above)
Comment (LeonidLapshin) [reply images/linux/scripts/installers/dotnetcore-sdk.sh:63]: as @al-cheb mentioned, these is no child processes error handling so I will use parallel tool as it has it as build-in option

File: images/linux/scripts/installers/dotnetcore-sdk.sh [hunk 4]
```diff
@@ -46,34 +46,32 @@ done
...
-    rsync -qav "$dest/sdk/" /usr/share/dotnet/sdk/
-    rm -rf "$dest"
-    rm "$tarball"
-done
-rm urls
+parallel --jobs 0 --halt soon,fail=1 \
+    'url="https://dotnetcli.blob.core.windows.net/dotnet/Sdk/{}/dotnet-sdk-{}-linux-x64.tar.gz"; \
+    download_with_retries $url' ::: "${sortedSdks[@]}"
+
+find . -name "*.tar.gz" | parallel --halt soon,fail=1 \
```
Comment (al-cheb) [inline images/linux/scripts/installers/dotnetcore-sdk.sh:65]: Why don't create a simple function and push array as  you are doing above for `download_with_retries` ? Current syntax is awkward and unreadable.

File: images/linux/scripts/installers/dotnetcore-sdk.sh [hunk 5]
```diff
@@ -46,34 +46,32 @@ done
...
+    'set -e
+    dest="./tmp-$(basename -s .tar.gz {})"; \
+    echo "Extracting {} to $dest"; \
+    mkdir "$dest" && tar -C "$dest" -xzf "{}"; \
+    rsync -qav "$dest/shared/" /usr/share/dotnet/shared/; \
```
Comment (al-cheb) [inline images/linux/scripts/installers/dotnetcore-sdk.sh:70]: Can we move a folder instead of copy the content?

Comment (LeonidLapshin) [reply images/linux/scripts/installers/dotnetcore-sdk.sh:70]: Checked these folders, it seems that this stage need merge archive's content is different, so rsync is used, and it seems that rsync use is better than mv in this particular case

File: images/linux/scripts/installers/dotnetcore-sdk.sh (hunk 4 above)
Comment (LeonidLapshin) [reply images/linux/scripts/installers/dotnetcore-sdk.sh]: Moved scriptblock to function, thank you!

File: images/linux/scripts/installers/dotnetcore-sdk.sh (hunk 5 above)
Comment (al-cheb) [reply images/linux/scripts/installers/dotnetcore-sdk.sh]: This stage doesn't need to merge anything, just move folder A-> B. And a move operation in the same partition will be faster than copy, plus the `rm -rf "$dest" "$ARCHIVE_NAME"` will be faster due to delete less files.

Comment (LeonidLapshin) [reply images/linux/scripts/installers/dotnetcore-sdk.sh]: ok, it will be faster, true :)

**Review by AlenaSviridenko — APPROVED** [review]

File: images/linux/scripts/installers/dotnetcore-sdk.sh [hunk 6]
```diff
@@ -46,34 +46,35 @@ done
...
     sdks=("${sdks[@]}" $(echo "${releases}" | jq '.releases[]' | jq '.sdk.version'))
     sdks=("${sdks[@]}" $(echo "${releases}" | jq '.releases[]' | jq '.sdks[]?' | jq '.version'))
+    rm ./${version}.json
 done
 
 sortedSdks=$(echo ${sdks[@]} | tr ' ' '\n' | grep -v preview | grep -v rc | grep -v display | cut -d\" -f2 | sort -u -r)
+extract_dotnet_sdk() {
+    local ARCHIVE_NAME="$1"
+    set -e
+    dest="./tmp-$(basename -s .tar.gz $ARCHIVE_NAME)"; \
```
Comment (al-cheb) [inline images/linux/scripts/installers/dotnetcore-sdk.sh:60]: Why do we need `; \` at end of lines?

Comment (LeonidLapshin) [reply images/linux/scripts/installers/dotnetcore-sdk.sh:60]: missed that, i'll remove this, thank you!

**Review by al-cheb — APPROVED** [review]

**Review by maxim-lobanov — APPROVED** [review]

**Comment by maxim-lobanov** [comment]
@bishal-pdMSFT could you please review this .NET Core related PR when you have a chance?

**Review by dmitry-shibanov — APPROVED** [review]

File: images/linux/scripts/installers/dotnetcore-sdk.sh [hunk 7]
```diff
@@ -46,34 +46,35 @@ done
...
 done
 
 sortedSdks=$(echo ${sdks[@]} | tr ' ' '\n' | grep -v preview | grep -v rc | grep -v display | cut -d\" -f2 | sort -u -r)
+extract_dotnet_sdk() {
+    local ARCHIVE_NAME="$1"
+    set -e
+    dest="./tmp-$(basename -s .tar.gz $ARCHIVE_NAME)"
+    echo "Extracting $ARCHIVE_NAME to $dest"
+    mkdir "$dest" && tar -C "$dest" -xzf "$ARCHIVE_NAME"
+    rsync -qav --remove-source-files "$dest/shared/" /usr/share/dotnet/shared/
```
Comment (miketimofeev) [inline images/linux/scripts/installers/dotnetcore-sdk.sh:63]: do we really need `rsync` or `mv` is fine too?

**Review by miketimofeev — APPROVED** [review]

**Comment by maxim-lobanov** [comment]
@bishal-pdMSFT could you please review this .NET Core related PR when you have a chance?

**Review by Darleev — APPROVED** [review]

**Pull request merged**